File: frontend/src/domain/research/extraction-queue.ts
import type { SourceId } from '../shared/identifier.ts'
import { isDropped, type SourceSummary } from './document.ts'

/** How one document's most recent extraction went.
 *
 * `detail` is only ever set on a failure, and it is the *only* account of that
 * failure anywhere: nothing durable records that an extraction was requested,
 * so a failed one appends nothing to any stream. See `finished()` in
 * `extraction_queue.py`. Losing it here means the reader is told a document
 * is not extracted and never told why.
 *
 * `entities`/`relationships` are the other way round -- present on a success,
 * absent on a failure -- and are reported rather than stored, so a reader can
 * tell "extracted, found nothing" from "extracted, found the graph".
 */
export interface ExtractionOutcome {
  readonly sourceId: SourceId
  readonly status: 'done' | 'failed'
  readonly detail: string | null
  readonly entities: number | null
  readonly relationships: number | null
}

/** The whole project's extraction queue, as `/sources/extraction-queue`
 *  answers it: one running item at most, the rest waiting, and the last
 *  outcome per document.
 *
 * Deliberately the same three-list shape as `DispatchBoard`, because it is the
 * same kind of thing read the same way -- one catch-up route rather than a
 * frame per document. The queue publishes no frames of its own at all (see
 * `get_extraction_queue`), which is why this read exists and why it is the
 * only way to learn that something is *waiting*.
 */
export interface ExtractionQueueBoard {
  readonly running: SourceId | null
  readonly queued: readonly SourceId[]
  readonly finished: readonly ExtractionOutcome[]
}

export const emptyExtractionQueue: ExtractionQueueBoard = {
  running: null,
  queued: [],
  finished: [],
}

/** What a document row should say about extraction, and whether it may offer
 *  it.
 *
 * One value rather than a handful of booleans on the row, because the states
 * are exclusive and a row that computed them separately would eventually draw
 * two at once -- "queued" beside "extracted" is the shape of that mistake.
 */
export type DocumentExtraction =
  /** Dropped documents are never offered extraction, matching the server:
   *  extract-all excludes them, so a control here would offer an action the
   *  bulk path has already decided against. */
  | { readonly kind: 'dropped' }
  /** Media, which nothing extracts yet. Its own state rather than `idle`,
   *  because `idle` is what the extract control is offered on: the server's
   *  `_unextracted` counts `kind == "text"` rows only, so a media row left
   *  idle would both offer a press extract-all has already decided against
   *  and inflate the "Extract all (N)" estimate beside it. */
  | { readonly kind: 'unextractable' }
  | { readonly kind: 'running' }
  | { readonly kind: 'queued' }
  | { readonly kind: 'failed'; readonly detail: string | null }
  | { readonly kind: 'extracted' }
  | { readonly kind: 'idle' }

/** Whether this row's control should be pressable. Derived rather than stored,
 *  so "may I press it" and "what does it say" cannot disagree. */
export const canExtract = (state: DocumentExtraction): boolean =>
  state.kind === 'idle' || state.kind === 'failed'

/** The order of these tests is the design, and each one is load-bearing.
 *
 * Dropped first: it is a property of the document rather than of the queue,
 * and a dropped document that somehow reached the queue still must not offer
 * the control. Running and queued next, so a *re*-extraction of an already
 * extracted document reports what is happening now rather than how it stood
 * before -- the same last-write-wins rule `byTopic` documents for dispatches.
 * Failure before success, because a failure is the more recent account and is
 * the one with no other record of itself.
 *
 * A `done` outcome counts as extracted even when the row still says otherwise:
 * `extracted` comes from the corpus projection and the queue answers from
 * memory, so between an extraction finishing and its projection catching up
 * the flag is stale by a moment. Trusting only the flag would offer "Extract"
 * on a document that had just been extracted.
 */
export const documentExtraction = (
  document: SourceSummary,
  board: ExtractionQueueBoard,
): DocumentExtraction => {
  if (isDropped(document)) return { kind: 'dropped' }
  // Second, and before the queue is consulted at all. A medium *does* reach
  // the queue now -- `perceive` enqueues under the medium's own source id --
  // but it reaches it to be perceived, never to be extracted, and the board
  // is one board with no field distinguishing the two. So a state read out of
  // it for a medium would be a transcription reported as "Extracting...",
  // which is the wrong account of what is happening and offers a control that
  // does nothing. The derived text source queues on its own and is where the
  // extraction progress belongs.
  if (document.kind === 'media') return { kind: 'unextractable' }
  if (board.running === document.sourceId) return { kind: 'running' }
  if (board.queued.includes(document.sourceId)) return { kind: 'queued' }
  const outcome = board.finished.find((row) => row.sourceId === document.sourceId)
  if (outcome?.status === 'failed') return { kind: 'failed', detail: outcome.detail }
  if (document.extracted || outcome?.status === 'done') return { kind: 'extracted' }
  return { kind: 'idle' }
}

/** What a media row should say about its *transcription*, which is a different
 *  question from `DocumentExtraction` over the same board.
 *
 * The third state B94 records as missing. `documentExtraction` answers
 * `unextractable` for every medium and must keep doing so -- a medium is not a
 * document and must not be offered to the extraction queue -- and `perceiveBusy`
 * is only true while the POST itself is in flight, which is milliseconds
 * against the minutes an hour of audio takes. Between those two the row showed
 * a live "Transcribe" button and nothing else.
 *
 * Read off the same board, keyed by the *medium's* id: `perceive` enqueues
 * under the medium and extraction under the derived id (`<id>#perceived`), so
 * the two share one queue and cannot collide on one key. See
 * `ExtractionQueue._drain`, which is where perception's outcome is recorded
 * with no entity counts.
 */
export type MediaPerception =
  | { readonly kind: 'transcribing' }
  | { readonly kind: 'queued' }
  | { readonly kind: 'failed'; readonly detail: string | null }
  /** Nothing to say: a text row, a dropped medium, or one the queue has never
   *  held. Deliberately *not* a `transcribed` state -- a finished perception is
   *  already reported by the row swapping its press for a "Transcript" link off
   *  `derivedSources`, and a second account of the same fact is how a row ends
   *  up drawing both. */
  | { readonly kind: 'idle' }

/** Order matters here exactly as it does in `documentExtraction`: running and
 *  queued before any past outcome, so a *re*-transcription reports what is
 *  happening now rather than how the last one went.
 *
 * Non-media rows answer `idle` rather than being refused, so a caller cannot
 * be wrong about which rows may ask. A text source can reach this queue on its
 * own account -- that is what extraction is -- and answering its extraction as
 * a transcription would be the mirror image of the mistake
 * `documentExtraction` guards against, so the kind is checked first and
 * nothing else is consulted.
 */
export const mediaPerception = (
  document: SourceSummary,
  board: ExtractionQueueBoard,
): MediaPerception => {
  if (document.kind !== 'media') return { kind: 'idle' }
  if (board.running === document.sourceId) return { kind: 'transcribing' }
  if (board.queued.includes(document.sourceId)) return { kind: 'queued' }
  const outcome = board.finished.find((row) => row.sourceId === document.sourceId)
  if (outcome?.status === 'failed') return { kind: 'failed', detail: outcome.detail }
  return { kind: 'idle' }
}

/** Whether a transcription press would do anything. Derived for
 *  `canExtract`'s reason: "may I press it" and "what does it say" cannot be
 *  allowed to disagree. */
export const canPerceive = (state: MediaPerception): boolean =>
  state.kind === 'idle' || state.kind === 'failed'

/** How many documents "extract all unextracted" would actually take on.
 *
 * Counted here rather than read from the server so the header can say a number
 * before anything is pressed. It is an estimate of the server's own set and is
 * allowed to differ: the server recomputes at press time, and the honest
 * report afterwards is the count the 202 returns, not this one.
 */
export const unextractedCount = (
  documents: readonly SourceSummary[],
  board: ExtractionQueueBoard,
): number => documents.filter((document) => canExtract(documentExtraction(document, board))).length

File: research_team/interfaces/web/extraction_queue.py
"""What has been asked for extraction, and what is waiting its turn.

`DispatchQueue` is the shape, and the reasoning in its module docstring
applies here almost line for line: a control on every row, FIFO rather than a
refusal, at most one in flight per project, drained by one background task,
provisional and never durable. Read that first; this records only where the
two differ, and both differences are the same difference -- **extraction
already has a progress channel and dispatch never did.**

**No feed, no frame type, no SSE pump.** `ExtractionActivity` carries
`storing` / `extracting` / `consolidating` / `consolidated` / `failed` to the
browser already, with its own listeners and its own catch-up route, and the
running item here is the one those frames describe. A second channel
publishing "this extraction is running" beside the one already saying
`extracting, chunk 3 of 40` would be two accounts of one thing, and the way
that goes wrong is not a crash -- it is a pane and a row disagreeing, with
nothing in either signature to catch it. So this queue publishes nothing. It
is read through its catch-up route, which the client refreshes when an
extraction frame arrives on the connection it already has.

The gap that leaves is real and is accepted: **a queue that changes with no
extraction frame behind it is not pushed anywhere.** Another tab queueing six
documents does not move this tab's rows until something else refreshes them.
That is a second-tab-only staleness in a local single-user tool, and the fix
-- a frame type, a pump, a `decodeFrame` case and a store -- costs more than
it buys until somebody is actually running two tabs against one project.

**The same document is never queued twice.** `DispatchQueue` deliberately does
not deduplicate, because dispatching a topic twice asks for two pieces of
work. Extracting one document twice asks for the same graph to be written
twice, at model cost, for no difference in the result. It matters most for
"extract all": pressed twice while the first pass is draining, a queue without
this would hold every document twice over.
"""

import asyncio
from collections import deque
from collections.abc import Awaitable, Callable
from contextlib import suppress
from dataclasses import dataclass
from typing import Any
from uuid import UUID

from research_team.application.knowledge import IngestReport


@dataclass
class _Pending:
    """One document waiting to be extracted."""

    source_id: str
    run: Callable[[], Awaitable[IngestReport | None]]


class ExtractionQueue:
    """One project's extractions: the running one, the waiting ones, the last ones."""

    def __init__(self) -> None:
        self._pending: dict[UUID, deque[_Pending]] = {}
        self._running: dict[UUID, str] = {}
        self._finished: dict[UUID, dict[str, dict[str, Any]]] = {}
        self._tasks: dict[UUID, asyncio.Task] = {}
        self._draining: set[UUID] = set()

    # ---------------- what the routes drive ----------------

    def start(
        self,
        project_id: UUID,
        source_id: str,
        run: Callable[[], Awaitable[IngestReport | None]],
    ) -> bool:
        """Enqueue one document. False if it was already queued or running.

        A bool rather than a frame, unlike `DispatchQueue.start`: there is no
        frame to hand back, because the caller's next read is the catch-up
        route and the running item's account comes from `ExtractionActivity`.
        What the route does need to know is whether this press changed
        anything, so "extract all" can answer how many it actually took on.

        `run` is a factory rather than a coroutine already in hand, matching
        `DispatchQueue.start`: an item that sits in a deque for a minute would
        otherwise be a live coroutine nobody has awaited, and one dropped by
        `cancel` would be one nobody ever will.

        `_draining` is maintained synchronously rather than checked through
        `task.done()`, for the reason `DispatchQueue` gives: `done()` is false
        for the whole window between the last item finishing and the drain
        coroutine returning, and a press landing there would queue behind a
        task that was already leaving.
        """
        if self._holds(project_id, source_id):
            return False
        self._pending.setdefault(project_id, deque()).append(_Pending(source_id, run))

        if project_id not in self._draining:
            self._draining.add(project_id)
            self._tasks[project_id] = asyncio.ensure_future(self._drain(project_id))
        return True

    def cancel(self, project_id: UUID) -> int:
        """Stop what is running and drop everything waiting. Returns how many went.

        Per project rather than per document, matching `DispatchQueue.cancel`
        and for the same reason: the UI offers one stop control, not one per
        queued row, so this does not have to honour an action nothing offers.

        A cancelled extraction leaves no `failed` note behind, because the
        `CancelledError` is raised at an `await` inside the ingest rather than
        being reported through it. The row simply stops being queued, which is
        what a person who just pressed stop expects to see.
        """
        waiting = self._pending.pop(project_id, deque())
        # The running *document*, not the drain task. `DispatchQueue` counts
        # the task, which is right there because it pops its item off the deque
        # before awaiting; here the task is scheduled by `start` and may not
        # have claimed anything yet, so counting it would report one more
        # document stopped than the project has -- two documents queued and
        # cancelled before the first turn of the loop would answer 3.
        stopped = len(waiting) + (1 if project_id in self._running else 0)

        task = self._tasks.get(project_id)
        if task is not None and not task.done():
            # Cancelled regardless of whether it had claimed an item: an
            # unclaimed drain task left running would find an empty deque and
            # return, but one that claims its item in the same turn this
            # returns would extract a document the caller just stopped.
            task.cancel()
        return stopped

    # ---------------- what the catch-up route reads ----------------

    def current(self, project_id: UUID) -> str | None:
        """The document being extracted, for a tab that arrived mid-run."""
        return self._running.get(project_id)

    def queued(self, project_id: UUID) -> tuple[str, ...]:
        """Every document waiting, in the order it will run."""
        return tuple(pending.source_id for pending in self._pending.get(project_id, ()))

    def finished(self, project_id: UUID) -> list[dict[str, Any]]:
        """How each document's most recent extraction went.

        Kept rather than cleared on completion, for `ExtractionActivity.last`'s
        reason: nothing durable records that an extraction was *requested*, so
        a failure has no other account of itself. `DocumentExtracted` records
        the successes, which is why a row that succeeded can also be read from
        the corpus projection -- but a failure appends nothing anywhere.
        """
        return list(self._finished.get(project_id, {}).values())

    async def wait(self, project_id: UUID) -> None:
        """Block until this project's queue drains. For tests, not routes."""
        task = self._tasks.get(project_id)
        if task is not None:
            # A cancelled drain is the expected end of `cancel`, not a failure
            # for a waiter to re-raise at its own call site.
            with suppress(asyncio.CancelledError):
                await task

    # ---------------- internals ----------------

    def _holds(self, project_id: UUID, source_id: str) -> bool:
        if self._running.get(project_id) == source_id:
            return True
        return any(
            pending.source_id == source_id for pending in self._pending.get(project_id, ())
        )

    async def _drain(self, project_id: UUID) -> None:
        """Extract this project's documents one at a time until none are waiting.

        **`None` from `run` means "done, with no counts to report", and the
        keys are omitted rather than zeroed.** Perception shares this queue --
        transcribing an hour of audio is the same kind of slow, and a second
        queue would be a second thing to cancel -- but it extracts nothing, and
        `entities: 0` on a finished transcription would read as "extraction
        found nothing" rather than "no extraction happened". The wire schema
        already treats both counts as optional (`extractionOutcomeDto`), so
        this needed no client change.

        A failure is recorded and the loop continues, matching
        `DispatchQueue._drain`: stopping on the first would turn one rate-limit
        refusal in the middle of "extract all" into forty documents silently
        skipped, and the log would say nothing, because a queued extraction
        appends no event of its own.
        """
        try:
            while True:
                waiting = self._pending.get(project_id) or deque()
                if not waiting:
                    return
                pending = waiting.popleft()
                self._running[project_id] = pending.source_id

                try:
                    report = await pending.run()
                except asyncio.CancelledError:
                    raise
                except Exception as error:  # noqa: BLE001 -- reported, not raised, from a task
                    outcome: dict[str, Any] = {
                        "source_id": pending.source_id,
                        "status": "failed",
                    }
                    outcome["detail"] = str(error)
                else:
                    outcome = {"source_id": pending.source_id, "status": "done"}
                    if report is not None:
                        outcome["entities"] = report.entity_count
                        outcome["relationships"] = report.relationship_count
                        # Reported unconditionally rather than only when
                        # non-zero. A key that appears only on a bad run is a
                        # key nobody builds a habit of reading, and the
                        # baseline is the half of this number that matters:
                        # `unresolved` alone says nothing, `unresolved`
                        # against `relationships` says whether the prompt is
                        # landing.
                        outcome["unresolved"] = report.unresolved_relationships
                        outcome["date_nodes"] = report.date_nodes
                        outcome["lifted_dates"] = report.lifted_dates

                self._finished.setdefault(project_id, {})[pending.source_id] = outcome
                self._running.pop(project_id, None)
        finally:
            # Unconditional, so a cancelled drain leaves the project
            # extractable. Without this, every later press would queue behind
            # a `_draining` entry belonging to a task that had gone.
            self._running.pop(project_id, None)
            self._draining.discard(project_id)
